src/crimson/os/seastore/cache.cc

```diff
@@ -357,65 +354,121 @@ void Cache::register_metrics()
-                 << " transactional efforts labeled by source, effort and counter";
-        std::ostringstream oss_metric;
-        oss_metric << category << "_efforts";
+    // invalidated efforts (non READ)
+    for (auto& [src, src_label] : labels_by_src) {
+      if (src == src_t::READ) {
+        // register src_t::READ later
+        continue;
+      }
+      auto& efforts = get_counter(stats.invalidated_efforts_by_src, src);
+      efforts = {};
+      for (auto& [effort_name, effort_label] : labels_by_effort) {
+        auto& effort = efforts.get_by_name(effort_name);
         metrics.add_group(
           "cache",
           {
             sm::make_counter(
-              oss_metric.str(),
-              value,
-              sm::description(oss_desc.str()),
-              {labels_by_src.find(src)->second,
-               labels_by_effort.find(effort_name)->second,
-               labels_by_counter.find(counter_name)->second}
+              "invalidated_extents",
+              effort.extents,
+              sm::description("extents of invalidated transactions"),
+              {src_label, effort_label}
+            ),
+            sm::make_counter(
+              "invalidated_extent_bytes",
+              effort.bytes,
+              sm::description("extent bytes of invalidated transactions"),
+              {src_label, effort_label}
             ),
           }
         );
-      };
-
-    auto get_efforts_by_category =
-      [this](const char* category) -> auto& {
-        if (strcmp(category, "committed") == 0) {
-          return stats.committed_efforts_by_src;
-        } else {
-          assert(strcmp(category, "invalidated") == 0);
-          return stats.invalidated_efforts_by_src;
-        }
-      };
+      } // effort_name
 
-    for (auto& category : {"committed", "invalidated"}) {
-      auto& efforts_by_src = get_efforts_by_category(category);
-      for (auto& [src, label] : labels_by_src) {
-        if (std::strcmp(category, "committed") == 0 && src == src_t::READ) {
-          // READ transaction won't commit
-          continue;
+      metrics.add_group(
+        "cache",
+        {
+          sm::make_counter(
+            "invalidated_delta_bytes",
+            efforts.mutate_delta_bytes,
+            sm::description("delta bytes of invalidated transactions"),
+            {src_label}
+          ),
         }
+      );
+    } // src
+
+    // invalidated efforts (READ)
+    // read transaction won't have non-read efforts
+    auto read_src_label = labels_by_src.find(src_t::READ)->second;
+    auto read_effort_label = labels_by_effort.find("READ")->second;
+    auto& read_efforts = get_counter(stats.invalidated_efforts_by_src, src_t::READ);
+    read_efforts = {};
+    metrics.add_group(
+      "cache",
+      {
+        sm::make_counter(
+          "invalidated_extents",
+          read_efforts.read.extents,
+          sm::description("extents of invalidated transactions"),
```

Comment:
read extents of invalidated read transactions?

Reply:
READ is already labeled by `read_src_label`, and I think our convention is not to encode label information into the description.

see https://github.com/ceph/ceph/pull/42261#discussion_r672006984